### PR TITLE
samx5x: don't assume that flash block size == flash region lock size

### PR DESCRIPTION
**Pull request by Qyriad**
This fixes flashing on samx5x chips with larger amounts of flash (e.g. 512KiB).

## Detailed description

Currently, we lock and unlock each flash region multiple times when the flash block size differs from the size of a protected flash region — i.e. the size of flash encoded by a single bit in the RUNLOCK register.

However, the RUNLOCK register also reads as "locked" while a write is still in progress, and we check that before each subsequent write. When we also manually lock after each block, and the block is smaller than a protected region, we misinterpret our own lock as an indication that the write is still in progress, and error out.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

### CODE REVIEW

File: src/target/samx5x.c [hunk 1]
```diff
@@ -474,11 +474,16 @@ static bool samx5x_flash_erase(target_flash_s *f, target_addr_t addr, size_t len
...
 
+	bool is_first_section = true;
+
 	for (size_t offset = 0; offset < len; offset += f->blocksize) {
 		target_mem_write32(t, SAMX5X_NVMC_ADDRESS, addr + offset);
 
-		/* Unlock */
-		samx5x_unlock_current_address(t);
+		/* If we're about to touch a new flash region, unlock it. */
+		if (is_first_section || ((offset % lock_region_size) == 0)) {
```
Comment (dragonmux) [inline src/target/samx5x.c:483]: The parens around the second half of the check are not required, please drop them like as in:
```suggestion
		if (is_first_section || (offset % lock_region_size) == 0) {
```

They suppress the GCC warning for accidental assignment which we want here.

File: src/target/samx5x.c [hunk 2]
```diff
@@ -496,8 +501,11 @@ static bool samx5x_flash_erase(target_flash_s *f, target_addr_t addr, size_t len
 			return false;
 		}
 
-		/* Lock */
-		samx5x_lock_current_address(t);
+		/* If we've just finished writing to a flash region, lock it. */
+		size_t next_offset = offset + f->blocksize;
```
Comment (dragonmux) [inline src/target/samx5x.c:505]: This can be `const`, let's make it `const`.

File: src/target/samx5x.c [hunk 3]
```diff
@@ -496,8 +501,11 @@ static bool samx5x_flash_erase(target_flash_s *f, target_addr_t addr, size_t len
...
+		if ((next_offset % lock_region_size) == 0) {
```
Comment (dragonmux) [inline src/target/samx5x.c:506]: The braces here are superfluous, please remove them.

**Review by dragonmux — CHANGES_REQUESTED** [review]
Nicely caught and figured out! There are a couple of stylistic changes to be made please, and then we'll be happy to merge this.

**Comment by ktemkin** [comment]
Feel free to make whatever stylistic changes you want :) 

We won't have time to sit down with @Qyriad and update this for a bit. [Already had to debug my debugger; now I need to get back to debugging my app. ;)]

**Comment by dragonmux** [comment]
Ok, thank you - we'll get on that shortly then so we can get this merged :)

**Review by dragonmux — APPROVED** [review]
We'll merge this as soon as the builds complete - we kept the changes strictly to the review items

**Pull request merged**